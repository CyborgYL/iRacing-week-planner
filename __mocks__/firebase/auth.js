export function getAuth() {
  return {};
}

export async function signOut() { }

export const signInWithEmailAndPassword = jest.fn(async () => {});

export async function createUserWithEmailAndPassword() { }

const authStateListeners = [];

export function onAuthStateChanged(auth, listener) {
  authStateListeners.push(listener);
}

export function testDispatchOnAuthStateChanged(user) {
  authStateListeners.forEach((listener) => {
    listener(user);
  });
}

export async function sendPasswordResetEmail() { }
