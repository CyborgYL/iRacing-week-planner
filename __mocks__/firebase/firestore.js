export const getFirestore = jest.fn(() => ({}));

export const collection = jest.fn(() => '654321');

export const doc = jest.fn(() => '123456');

export const getDoc = jest.fn(async () => ({}));

export const setDoc = jest.fn(async () => {});
