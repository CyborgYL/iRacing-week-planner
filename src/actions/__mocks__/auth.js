export const LOADING_AUTH = 'AUTH/LOADING_SIGN_IN';
export const ERROR_AUTH = 'AUTH/ERROR_AUTH';
export const ERROR_RESET = 'AUTH/ERROR_RESET';
export const ERROR_ACKNOWLEDGE = 'AUTH/ERROR_ACKNOWLEDGE';
export const SIGNED_IN = 'AUTH/SIGNED_IN';
export const SIGNED_OUT = 'AUTH/SIGNED_OUT';
export const RESET_SENT = 'AUTH/RESET_SENT';
export const LOADING_RESET = 'AUTH/LOADING_RESET';

export function signOut() {
  return async (dispatch) => {
    dispatch({ type: LOADING_AUTH });
    dispatch({ type: SIGNED_OUT });
  };
}

export function signIn() {
  return async (dispatch) => {
    dispatch({ type: LOADING_AUTH });
  };
}

export function createAccount() {
  return async (dispatch) => {
    dispatch({ type: LOADING_AUTH });
  };
}

export function acknowledgeAuthError() {
  return { type: ERROR_ACKNOWLEDGE };
}

export function signedIn(user) {
  return async (dispatch) => {
    await dispatch({ type: SIGNED_IN, user });
  };
}

export function startListener() {
  return async (dispatch) => {
    dispatch(signedIn(null));
  };
}

export function forgottenPassword() {
  return async (dispatch) => {
    dispatch({ type: LOADING_RESET });
    dispatch({ type: RESET_SENT });
  };
}
