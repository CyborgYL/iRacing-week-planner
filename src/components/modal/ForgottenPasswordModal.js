import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import {
  acknowledgeAuthError as acknowledgeAuthErrorAction,
  ERROR_RESET,
  forgottenPassword as forgottenPasswordAction,
} from '../../actions/auth';
import Modal from './Modal';
import styles from '../../styles/main.module.scss';

function ForgottenPasswordModal({ isOpen, onClose, error, loading, forgottenPassword, acknowledgeAuthError }) {
  const { t } = useTranslation();
  const [email, setEmail] = useState('');
  const [thanksModal, setThanksModal] = useState(false);

  const submitClick = async () => {
    const output = await forgottenPassword(email);
    if (output.type !== ERROR_RESET) {
      setEmail('');
      setThanksModal(true);
    }
  };

  const closeThanks = () => {
    setThanksModal(false);
    onClose();
  };

  return (
    <Modal isOpen={isOpen} onClose={onClose} title={t('Reset Password')} doneAction={onClose} showFooter={false}>
      <div className={styles['container-fluid']}>
        <p>{t('Enter your email address to reset your password.')}</p>
        {loading ? (
          <div>{t('Loading')}</div>
        ) : (
          <form onSubmit={(e) => {
            e.preventDefault();
            submitClick();
          }}
          >
            {error && (
              <div className={`${styles.alert} ${styles['alert-warning']} ${styles['alert-dismissible']}`} role="alert">
                <button className={styles.close} type="button" aria-label="Close" onClick={acknowledgeAuthError}>
                  <span
                    aria-hidden="true"
                  >
                    &times;
                  </span>
                </button>
                {t(error.message)}
              </div>
            )}
            <div className={styles['form-group']}>
              <label htmlFor="loginEmail">{t('Email address')}</label>
              <input
                type="email"
                className={styles['form-control']}
                id="loginEmail"
                placeholder={t('Email address')}
                onChange={(e) => setEmail(e.target.value)}
                value={email}
              />
            </div>
            <button type="submit" className={`${styles.btn} ${styles['btn-default']}`}>{t('Submit')}</button>
          </form>
        )}
      </div>
      <Modal
        isOpen={thanksModal}
        onClose={closeThanks}
        title={t('Reset Password')}
        doneAction={closeThanks}
      >
        <div className={styles['container-fluid']}>
          <p>{t('Thanks, please check your email for further details.')}</p>
        </div>
      </Modal>
    </Modal>
  );
}

ForgottenPasswordModal.propTypes = {
  onClose: PropTypes.func,
  isOpen: PropTypes.bool.isRequired,
  forgottenPassword: PropTypes.func.isRequired,
  acknowledgeAuthError: PropTypes.func.isRequired,
  error: PropTypes.shape({
    message: PropTypes.string,
  }),
  loading: PropTypes.bool.isRequired,
};

ForgottenPasswordModal.defaultProps = {
  onClose: () => {},
  isOpen: false,
  error: null,
};

const mapStateToProps = (state) => ({
  error: state.auth.errorReset,
  loading: state.auth.loadingReset,
});

const mapDispatchToProps = (dispatch) => bindActionCreators({
  forgottenPassword: forgottenPasswordAction,
  acknowledgeAuthError: acknowledgeAuthErrorAction,
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(ForgottenPasswordModal);
