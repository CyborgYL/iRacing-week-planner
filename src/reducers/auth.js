// @flow

import { initializeApp } from 'firebase/app';
import type { FirebaseApp } from 'firebase/app';

import {
  LOADING_AUTH, SIGNED_OUT, SIGNED_IN, LOADING_RESET, RESET_SENT, ERROR_AUTH, ERROR_ACKNOWLEDGE, ERROR_RESET,
} from '../actions/auth';
import { firebaseConfig } from '../config';

const firebaseApp = initializeApp(firebaseConfig);

type AuthUser = {

};

type AuthState = {
  user: AuthUser,
  firebaseApp: FirebaseApp,
  loadingAuth: boolean,
  loadingReset: boolean,
  errorAuth: { message: string } | null,
  errorReset: { message: string } | null,
};

type AuthReducerArgs = {
  type: string,
  user?: AuthUser,
  error?: { message: string },
};

export default function authReducer(
  state: AuthState = {
    user: null,
    firebaseApp,
    loadingAuth: true,
    loadingReset: false,
    errorAuth: null,
    errorReset: null,
  },
  { type, user, error }: AuthReducerArgs,
): AuthState {
  if (type === SIGNED_OUT) {
    return { ...state, user: null };
  }

  if (type === LOADING_AUTH) {
    return { ...state, loadingAuth: true };
  }

  if (type === ERROR_AUTH) {
    return { ...state, loadingAuth: false, errorAuth: error };
  }

  if (type === ERROR_RESET) {
    return { ...state, loadingReset: false, errorReset: error };
  }

  if (type === ERROR_ACKNOWLEDGE) {
    return { ...state, errorAuth: null, errorReset: null };
  }

  if (type === SIGNED_IN) {
    return { ...state, loadingAuth: false, errorAuth: null, user };
  }

  if (type === LOADING_RESET) {
    return { ...state, loadingReset: true };
  }

  if (type === RESET_SENT) {
    return { ...state, loadingReset: false, errorReset: null };
  }

  return state;
}
