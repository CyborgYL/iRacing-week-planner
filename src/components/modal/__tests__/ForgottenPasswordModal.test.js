import { describe, test } from '@jest/globals';
import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { getAuth, sendPasswordResetEmail } from 'firebase/auth';

import flushPromises from 'flush-promises';
import ForgottenPasswordModal from '../ForgottenPasswordModal';
import { ERROR_ACKNOWLEDGE } from '../../../actions/auth';

jest.mock('firebase/auth');
jest.useFakeTimers();

describe('components/modal/ForgottenPasswordModal', () => {
  const mockStore = configureMockStore([thunk]);

  test('renders closed', async () => {
    const store = mockStore({ auth: { errorReset: null, loadingReset: false } });
    const component = renderer.create(<Provider store={store}><ForgottenPasswordModal isOpen={false} /></Provider>);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('happy path', async () => {
    const firebaseAuth = {};
    getAuth.mockReturnValue(firebaseAuth);
    const store = mockStore({ auth: { errorReset: null, loadingReset: false, firebaseApp: {} } });
    let component;

    await act(async () => {
      component = renderer.create(<Provider store={store}><ForgottenPasswordModal isOpen /></Provider>);
    });

    expect(component.toJSON()).toMatchSnapshot();

    await act(async () => {
      await component.root.findByProps({ id: 'loginEmail' }).props.onChange({ target: { value: 'tom@example.com' } });
    });

    await act(async () => {
      await component.root.findByType('form').props.onSubmit({ preventDefault: () => {} });
    });

    expect(sendPasswordResetEmail).toHaveBeenCalledWith(firebaseAuth, 'tom@example.com');

    jest.runAllTimers();
    await flushPromises();

    expect(component.toJSON()).toMatchSnapshot();

    await act(async () => {
      await component.root
        .findByProps({ id: 'forgottenPasswordThanks' })
        .findByProps({ 'aria-label': 'Close' })
        .props.onClick({ preventDefault: () => {}, stopPropagation: () => {} });
    });

    jest.runAllTimers();
    await flushPromises();

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('reset error', async () => {
    const firebaseAuth = {};
    getAuth.mockReturnValue(firebaseAuth);
    sendPasswordResetEmail.mockRejectedValueOnce({ message: 'error' });

    const store = mockStore({ auth: { errorReset: null, loadingReset: false, firebaseApp: {} } });
    let component;

    await act(async () => {
      component = renderer.create(<Provider store={store}><ForgottenPasswordModal isOpen /></Provider>);
    });

    await act(async () => {
      await component.root.findByType('form').props.onSubmit({ preventDefault: () => {} });
    });

    jest.runAllTimers();
    await flushPromises();

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('renders error', async () => {
    const store = mockStore({ auth: { errorReset: { message: 'Error message' }, loadingReset: false } });
    const dispatchSpy = jest.spyOn(store, 'dispatch');
    let component;

    await act(async () => {
      component = renderer.create(<Provider store={store}><ForgottenPasswordModal isOpen /></Provider>);
    });

    expect(component.toJSON()).toMatchSnapshot();

    await act(async () => {
      await component.root
        .findByProps({ role: 'alert' })
        .findByProps({ 'aria-label': 'Close' }).props.onClick();
    });

    expect(dispatchSpy).toHaveBeenCalledWith({ type: ERROR_ACKNOWLEDGE });
  });

  test('renders loading', async () => {
    const store = mockStore({ auth: { errorReset: null, loadingReset: true } });
    const component = renderer.create(<Provider store={store}><ForgottenPasswordModal /></Provider>);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
