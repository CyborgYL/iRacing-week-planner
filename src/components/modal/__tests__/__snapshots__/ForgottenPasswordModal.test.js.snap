// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/modal/ForgottenPasswordModal happy path 1`] = `
<div
  className="ReactModal__Overlay"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(255, 255, 255, 0.75)",
      "bottom": 0,
      "left": 0,
      "position": "fixed",
      "right": 0,
      "top": 0,
    }
  }
>
  <div
    aria-modal={true}
    className="ReactModal__Content Modal__Bootstrap modal-dialog modal-lg"
    id="forgottenPasswordModal"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    role="dialog"
    style={Object {}}
    tabIndex="-1"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <button
          aria-label="Close"
          className="close"
          onClick={[Function]}
          type="button"
        >
          <span
            className="glyphicon glyphicon-remove"
          />
        </button>
        <h4
          className="modal-title"
        >
          Reset Password
        </h4>
      </div>
      <div
        className="modal-body"
        style={
          Object {
            "maxHeight": "55vh",
            "overflowY": "auto",
          }
        }
      >
        <div
          className="container-fluid"
        >
          <p>
            Enter your email address to reset your password.
          </p>
          <form
            onSubmit={[Function]}
          >
            <div
              className="form-group"
            >
              <label
                htmlFor="loginEmail"
              >
                Email address
              </label>
              <input
                className="form-control"
                id="loginEmail"
                onChange={[Function]}
                placeholder="Email address"
                type="email"
                value=""
              />
            </div>
            <button
              className="btn btn-default"
              type="submit"
            >
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/modal/ForgottenPasswordModal happy path 2`] = `
<div
  className="ReactModal__Overlay ReactModal__Overlay--after-open"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(255, 255, 255, 0.75)",
      "bottom": 0,
      "left": 0,
      "position": "fixed",
      "right": 0,
      "top": 0,
    }
  }
>
  <div
    aria-modal={true}
    className="ReactModal__Content ReactModal__Content--after-open Modal__Bootstrap modal-dialog modal-lg"
    id="forgottenPasswordModal"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    role="dialog"
    style={Object {}}
    tabIndex="-1"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <button
          aria-label="Close"
          className="close"
          onClick={[Function]}
          type="button"
        >
          <span
            className="glyphicon glyphicon-remove"
          />
        </button>
        <h4
          className="modal-title"
        >
          Reset Password
        </h4>
      </div>
      <div
        className="modal-body"
        style={
          Object {
            "maxHeight": "55vh",
            "overflowY": "auto",
          }
        }
      >
        <div
          className="container-fluid"
        >
          <p>
            Enter your email address to reset your password.
          </p>
          <form
            onSubmit={[Function]}
          >
            <div
              className="form-group"
            >
              <label
                htmlFor="loginEmail"
              >
                Email address
              </label>
              <input
                className="form-control"
                id="loginEmail"
                onChange={[Function]}
                placeholder="Email address"
                type="email"
                value=""
              />
            </div>
            <button
              className="btn btn-default"
              type="submit"
            >
              Submit
            </button>
          </form>
        </div>
        <div
          className="ReactModal__Overlay ReactModal__Overlay--after-open"
          onClick={[Function]}
          onMouseDown={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(255, 255, 255, 0.75)",
              "bottom": 0,
              "left": 0,
              "position": "fixed",
              "right": 0,
              "top": 0,
            }
          }
        >
          <div
            aria-modal={true}
            className="ReactModal__Content ReactModal__Content--after-open Modal__Bootstrap modal-dialog modal-lg"
            id="forgottenPasswordThanks"
            onClick={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            role="dialog"
            style={Object {}}
            tabIndex="-1"
          >
            <div
              className="modal-content"
            >
              <div
                className="modal-header"
              >
                <button
                  aria-label="Close"
                  className="close"
                  onClick={[Function]}
                  type="button"
                >
                  <span
                    className="glyphicon glyphicon-remove"
                  />
                </button>
                <h4
                  className="modal-title"
                >
                  Reset Password
                </h4>
              </div>
              <div
                className="modal-body"
                style={
                  Object {
                    "maxHeight": "55vh",
                    "overflowY": "auto",
                  }
                }
              >
                <div
                  className="container-fluid"
                >
                  <p>
                    Thanks, please check your email for further details.
                  </p>
                </div>
              </div>
              <div
                className="modal-footer"
              >
                <button
                  className="btn btn-primary"
                  onClick={[Function]}
                  type="button"
                >
                  Close
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/modal/ForgottenPasswordModal happy path 3`] = `
<div
  className="ReactModal__Overlay ReactModal__Overlay--after-open"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(255, 255, 255, 0.75)",
      "bottom": 0,
      "left": 0,
      "position": "fixed",
      "right": 0,
      "top": 0,
    }
  }
>
  <div
    aria-modal={true}
    className="ReactModal__Content ReactModal__Content--after-open Modal__Bootstrap modal-dialog modal-lg"
    id="forgottenPasswordModal"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    role="dialog"
    style={Object {}}
    tabIndex="-1"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <button
          aria-label="Close"
          className="close"
          onClick={[Function]}
          type="button"
        >
          <span
            className="glyphicon glyphicon-remove"
          />
        </button>
        <h4
          className="modal-title"
        >
          Reset Password
        </h4>
      </div>
      <div
        className="modal-body"
        style={
          Object {
            "maxHeight": "55vh",
            "overflowY": "auto",
          }
        }
      >
        <div
          className="container-fluid"
        >
          <p>
            Enter your email address to reset your password.
          </p>
          <form
            onSubmit={[Function]}
          >
            <div
              className="form-group"
            >
              <label
                htmlFor="loginEmail"
              >
                Email address
              </label>
              <input
                className="form-control"
                id="loginEmail"
                onChange={[Function]}
                placeholder="Email address"
                type="email"
                value=""
              />
            </div>
            <button
              className="btn btn-default"
              type="submit"
            >
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/modal/ForgottenPasswordModal renders closed 1`] = `null`;

exports[`components/modal/ForgottenPasswordModal renders error 1`] = `
<div
  className="ReactModal__Overlay"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(255, 255, 255, 0.75)",
      "bottom": 0,
      "left": 0,
      "position": "fixed",
      "right": 0,
      "top": 0,
    }
  }
>
  <div
    aria-modal={true}
    className="ReactModal__Content Modal__Bootstrap modal-dialog modal-lg"
    id="forgottenPasswordModal"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    role="dialog"
    style={Object {}}
    tabIndex="-1"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <button
          aria-label="Close"
          className="close"
          onClick={[Function]}
          type="button"
        >
          <span
            className="glyphicon glyphicon-remove"
          />
        </button>
        <h4
          className="modal-title"
        >
          Reset Password
        </h4>
      </div>
      <div
        className="modal-body"
        style={
          Object {
            "maxHeight": "55vh",
            "overflowY": "auto",
          }
        }
      >
        <div
          className="container-fluid"
        >
          <p>
            Enter your email address to reset your password.
          </p>
          <form
            onSubmit={[Function]}
          >
            <div
              className="alert alert-warning alert-dismissible"
              role="alert"
            >
              <button
                aria-label="Close"
                className="close"
                onClick={[Function]}
                type="button"
              >
                <span
                  aria-hidden="true"
                >
                  ×
                </span>
              </button>
              Error message
            </div>
            <div
              className="form-group"
            >
              <label
                htmlFor="loginEmail"
              >
                Email address
              </label>
              <input
                className="form-control"
                id="loginEmail"
                onChange={[Function]}
                placeholder="Email address"
                type="email"
                value=""
              />
            </div>
            <button
              className="btn btn-default"
              type="submit"
            >
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/modal/ForgottenPasswordModal renders loading 1`] = `null`;

exports[`components/modal/ForgottenPasswordModal reset error 1`] = `
<div
  className="ReactModal__Overlay ReactModal__Overlay--after-open"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(255, 255, 255, 0.75)",
      "bottom": 0,
      "left": 0,
      "position": "fixed",
      "right": 0,
      "top": 0,
    }
  }
>
  <div
    aria-modal={true}
    className="ReactModal__Content ReactModal__Content--after-open Modal__Bootstrap modal-dialog modal-lg"
    id="forgottenPasswordModal"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    role="dialog"
    style={Object {}}
    tabIndex="-1"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <button
          aria-label="Close"
          className="close"
          onClick={[Function]}
          type="button"
        >
          <span
            className="glyphicon glyphicon-remove"
          />
        </button>
        <h4
          className="modal-title"
        >
          Reset Password
        </h4>
      </div>
      <div
        className="modal-body"
        style={
          Object {
            "maxHeight": "55vh",
            "overflowY": "auto",
          }
        }
      >
        <div
          className="container-fluid"
        >
          <p>
            Enter your email address to reset your password.
          </p>
          <form
            onSubmit={[Function]}
          >
            <div
              className="form-group"
            >
              <label
                htmlFor="loginEmail"
              >
                Email address
              </label>
              <input
                className="form-control"
                id="loginEmail"
                onChange={[Function]}
                placeholder="Email address"
                type="email"
                value=""
              />
            </div>
            <button
              className="btn btn-default"
              type="submit"
            >
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
